internal/features/stacks/jobs/validation.go

```diff
@@ -0,0 +1,90 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package jobs
+
+import (
+	"context"
+	"path"
+
+	"github.com/hashicorp/hcl-lang/decoder"
+	"github.com/hashicorp/hcl-lang/lang"
+	"github.com/hashicorp/hcl/v2"
+	lsctx "github.com/hashicorp/terraform-ls/internal/context"
+	idecoder "github.com/hashicorp/terraform-ls/internal/decoder"
+	"github.com/hashicorp/terraform-ls/internal/document"
+	"github.com/hashicorp/terraform-ls/internal/features/stacks/ast"
+	stackDecoder "github.com/hashicorp/terraform-ls/internal/features/stacks/decoder"
+	"github.com/hashicorp/terraform-ls/internal/features/stacks/state"
+	"github.com/hashicorp/terraform-ls/internal/job"
+	lsp "github.com/hashicorp/terraform-ls/internal/lsp"
+	globalAst "github.com/hashicorp/terraform-ls/internal/terraform/ast"
+	"github.com/hashicorp/terraform-ls/internal/terraform/module/operation"
+)
+
+func SchemaStackValidation(ctx context.Context, stackStore *state.StackStore, moduleFeature stackDecoder.ModuleReader, stackPath string) error {
+	rpcContext := lsctx.DocumentContext(ctx)
+	isMatchingLanguageId := rpcContext.LanguageID == lsp.Stacks.String() || rpcContext.LanguageID == lsp.Deploy.String()
+	if !isMatchingLanguageId {
+		return nil
+	}
+
+	record, err := stackStore.StackRecordByPath(stackPath)
+	if err != nil {
+		return err
+	}
+
+	// Avoid validation if it is already in progress or already finished
+	if record.DiagnosticsState[globalAst.SchemaValidationSource] != operation.OpStateUnknown && !job.IgnoreState(ctx) {
+		return job.StateNotChangedErr{Dir: document.DirHandleFromPath(stackPath)}
+	}
+
+	err = stackStore.SetDiagnosticsState(stackPath, globalAst.SchemaValidationSource, operation.OpStateLoading)
+	if err != nil {
+		return err
+	}
+
+	d := decoder.NewDecoder(&stackDecoder.PathReader{
+		StateReader:  stackStore,
+		ModuleReader: moduleFeature,
+	})
+	d.SetContext(idecoder.DecoderContext(ctx))
+
+	decoder, err := d.Path(lang.Path{
+		Path:       stackPath,
+		LanguageID: rpcContext.LanguageID,
```

Comment:
I think we should not create the path context based on the language ID from rpcContext, but rather use a static value.

If we use the ID from the context, we would only be validating the current file and not the whole Stack. When you open a folder, you expect the whole Stack to be validated, so you don't have to open deployment files separately to see their diagnostics.

Reply:
What do you think about creating two separate path contexts and the combining their diagnostics? Something along the lines of

```go
d := decoder.NewDecoder(&fdecoder.PathReader{
	StateReader:  modStore,
	ModuleReader: moduleFeature,
})
d.SetContext(idecoder.DecoderContext(ctx))

diags := make(lang.DiagnosticsMap)

stackPathCtx, err := d.Path(lang.Path{
	Path:       modPath,
	LanguageID: ilsp.Stacks.String(),
})
if err != nil {
	return err
}
stacksDiags, err := stackPathCtx.Validate(ctx)
if err != nil {
	return err
}
diags = diags.Extend(stacksDiags)

deployPathCtx, err := d.Path(lang.Path{
	Path:       modPath,
	LanguageID: ilsp.Deploy.String(),
})
if err != nil {
	return err
}
deployDiags, err := deployPathCtx.Validate(ctx)
if err != nil {
	return err
}
diags = diags.Extend(deployDiags)
```

Reply:
I initially had it this way, then in an attempt to reduce duplication consolidated to one. I can see your point about the other types of change notifications (open, etc) requiring a validation of the entire stack and deploy files which would require having both decoders.

In the situation of a didChange event however, we know we're dealing with one file and do not have to walk everything all over again. So I kept the single decoder here.